File: src/Desktop.AI.App/Desktop.AI.App/Interop/TextToSpeech/TextToSpeech.ts
class TextToSpeech {

    speak(text: string, voiceUri: string = "") {
        return new Promise<void>((resolve) => {
            let voices = speechSynthesis.getVoices();
            let voice;
            if (voiceUri === "" && voices.filter(i => i.lang == "en-GB").length > 0) {
                voice = voices.filter(i => i.lang == "en-GB")[0];
            } else {
                voice = voiceUri === "" ? voices.filter(i => i.default)[0] : voices.filter(i => i.voiceURI === voiceUri)[0];
            }

            let utterance = new SpeechSynthesisUtterance();
            utterance.text = text;
            utterance.lang = "en"
            utterance.voice = voice;
            speechSynthesis.speak(utterance);
            utterance.addEventListener("end", () => {
                resolve();
            });
        });
    }

    cancel() {
        speechSynthesis.cancel();
    }

    getVoices() {
        return new Promise((resolve) => {

            let voices = speechSynthesis.getVoices();

            if (voices.length == 0) {
                speechSynthesis.onvoiceschanged = () => {

                    voices = speechSynthesis.getVoices();
                    let mappedVoices = voices.map((voice) => ({ Name: voice.name, Lang: voice.lang, VoiceURI: voice.voiceURI }));
                    resolve(mappedVoices);

                }
            } else {
                let mappedVoices = voices.map((voice) => ({ Name: voice.name, Lang: voice.lang, VoiceURI: voice.voiceURI }));
                resolve(mappedVoices);
            }

        });
    }
}

export function createTextToSpeech(): TextToSpeech {
    return new TextToSpeech();
}

File: src/Desktop.AI.App/Desktop.AI.App/Interop/TextToSpeech/TextToSpeech.cs
using Desktop.AI.App.Models;
using Microsoft.JSInterop;

namespace Desktop.AI.App.Interop.TextToSpeech
{
    public class TextToSpeech :  ITextToSpeech
    {
        private const string _modulePath = "./interop/TextToSpeech/TextToSpeech.js";
        private readonly IJSRuntime _jSRuntime;
        private IJSObjectReference _module = null!;
        private IJSObjectReference _textToSpeech = null!;

        public TextToSpeech(IJSRuntime jSRuntime)
        {
            _jSRuntime = jSRuntime;
        }

        public async Task Initialise()
        {
            _module = await _jSRuntime!.InvokeAsync<IJSObjectReference>("import", _modulePath);
            _textToSpeech = await _module.InvokeAsync<IJSObjectReference>("createTextToSpeech");
        }

        public async Task Speak(string text, string voiceUri, CancellationToken canellationToken)
        {
            try
            {
                await _textToSpeech.InvokeAsync<object>("speak", canellationToken, text, voiceUri);
            }
            catch (TaskCanceledException) { }
        }

        public async Task Cancel()
        {
            await _textToSpeech.InvokeVoidAsync("cancel");
        }

        public async Task<List<Voice>> GetVoices()
        {
            var voices = await _textToSpeech.InvokeAsync<List<Voice>>("getVoices", CancellationToken.None);
            return voices;
        }

        public async ValueTask DisposeAsync()
        {
            if (_module != null)
            {
                await _module.DisposeAsync();
            }

            if (_textToSpeech != null)
            {
                await _textToSpeech.DisposeAsync();
            }

            GC.SuppressFinalize(this);
        }
    }
}
